in-project stubs cause mypy to bail in 0.971

<!--
  If you're new to mypy and you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form
  instead.
  Please also consider:

  - checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
  - searching our issue tracker: https://github.com/python/mypy/issues to see if
    it's already been reported
  - asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

We have a project that includes type stubs for a python package that does not provide it's own typing. We followed guidance per https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-library-stubs-or-py-typed-marker number 3

```
(venv) vfazio@vfazio2 ~/development/xtf $ grep mypy_path mypy.ini
mypy_path=$MYPY_CONFIG_FILE_DIR/stubs
```

In mypy 0.961, type checks pass

```
(venv) vfazio@vfazio2 ~/development/xtf $ mypy -V; mypy xes/
mypy 0.961 (compiled: no)
Success: no issues found in 86 source files
```
In mypy 0.971, these now fail because the current path is being added to the path search 

```
(venv) vfazio@vfazio2 ~/development/xtf :( $ mypy -V; mypy xes/
mypy 0.971 (compiled: no)
['/mnt/development/mypy/venv/lib/python3.10/site-packages', '/mnt/development/xtf']
/mnt/development/xtf is in the MYPYPATH. Please remove it.
See https://mypy.readthedocs.io/en/stable/running_mypy.html#how-mypy-handles-imports for more info

```

If i modify `mypy/pyinfo.py::getsearchdirs` to exclude the current working directory (effectively reverting 2132036a019ed26d77389c0a9af9b928911a5846) the type checks "work" except I now get type failures for the stubs

```
(venv) vfazio@vfazio2 ~/development/xtf $ mypy -V; mypy xes/
mypy 0.971 (compiled: no)
stubs/redacted/path/file.pyi:49:9: error: Function is missing a return type annotation  [no-untyped-def]
stubs/redacted/path/file.pyi:49:9: note: Use "-> None" if function does not return a value
<snip>
Found 58 errors in 2 files (checked 86 source files)

```

<!--
Note: If the problem you are reporting is about a specific library function, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues
-->

**Expected Behavior**

<!--
  How did you expect your project to behave?
  It’s fine if you’re not sure your understanding is correct.
  Write down what you thought would happen. If you just expected no errors, you can delete this section.
-->

Failures wouldn't occur or documentation would be updated to reflect how to accommodate this use case.

**Actual Behavior**

<!--
  Did something go wrong?
  Is something broken, or not behaving as you expected?
-->

mypy now fails to type check our code because we include stubs as part of our source tree

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 0.971
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.10.4
- Operating system and version: Ubuntu 20.04

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -1051,11 +1051,10 @@ def process_options(args: List[str],
     # Set target.
     if special_opts.modules + special_opts.packages:
         options.build_type = BuildType.MODULE
-        search_dirs = get_search_dirs(options.python_executable)
-        search_paths = SearchPaths((os.getcwd(),),
-                                   tuple(mypy_path() + options.mypy_path),
-                                   tuple(search_dirs),
-                                   ())
+        sys_path, _ = get_search_dirs(options.python_executable)
+        search_paths = SearchPaths(
+            (os.getcwd(),), tuple(mypy_path() + options.mypy_path), tuple(sys_path), ()
+        )
         targets = []
         # TODO: use the same cache that the BuildManager will
         cache = FindModuleCache(search_paths, fscache, options)

--- a/mypy/modulefinder.py
+++ b/mypy/modulefinder.py
@@ -723,7 +723,7 @@ def default_lib_path(data_dir: str,
 
 
 @functools.lru_cache(maxsize=None)
-def get_search_dirs(python_executable: Optional[str]) -> List[str]:
+def get_search_dirs(python_executable: Optional[str]) -> Tuple[List[str], List[str]]:
     """Find package directories for given python.
 
     This runs a subprocess call, which generates a list of the directories in sys.path.
@@ -732,15 +732,15 @@ def get_search_dirs(python_executable: Optional[str]) -> List[str]:
     """
 
     if python_executable is None:
-        return []
+        return ([], [])
     elif python_executable == sys.executable:
         # Use running Python's package dirs
-        sys_path = pyinfo.getsearchdirs()
+        sys_path, site_packages = pyinfo.getsearchdirs()
     else:
         # Use subprocess to get the package directory of given Python
         # executable
         try:
-            sys_path = ast.literal_eval(
+            sys_path, site_packages = ast.literal_eval(
                 subprocess.check_output([python_executable, pyinfo.__file__, 'getsearchdirs'],
                 stderr=subprocess.PIPE).decode())
         except OSError as err:
@@ -748,7 +748,7 @@ def get_search_dirs(python_executable: Optional[str]) -> List[str]:
             raise CompileError(
                 [f"mypy: Invalid python executable '{python_executable}': {reason}"]
             ) from err
-    return sys_path
+    return sys_path, site_packages
 
 
 def add_py2_mypypath_entries(mypypath: List[str]) -> List[str]:
@@ -837,22 +837,26 @@ def compute_search_paths(sources: List[BuildSource],
     if options.python_version[0] == 2:
         mypypath = add_py2_mypypath_entries(mypypath)
 
-    search_dirs = get_search_dirs(options.python_executable)
-    for search_dir in search_dirs:
-        assert search_dir not in lib_path
-        if (search_dir in mypypath or
-                any(p.startswith(search_dir + os.path.sep) for p in mypypath) or
-                (os.path.altsep
-                    and any(p.startswith(search_dir + os.path.altsep) for p in mypypath))):
-            print(f"{search_dir} is in the MYPYPATH. Please remove it.", file=sys.stderr)
+    sys_path, site_packages = get_search_dirs(options.python_executable)
+    # We only use site packages for this check
+    for site in site_packages:
+        assert site not in lib_path
+        if (
+            site in mypypath
+            or any(p.startswith(site + os.path.sep) for p in mypypath)
+            or (os.path.altsep and any(p.startswith(site + os.path.altsep) for p in mypypath))
+        ):
+            print(f"{site} is in the MYPYPATH. Please remove it.", file=sys.stderr)
             print("See https://mypy.readthedocs.io/en/stable/running_mypy.html"
                   "#how-mypy-handles-imports for more info", file=sys.stderr)
             sys.exit(1)
 
-    return SearchPaths(python_path=tuple(reversed(python_path)),
-                       mypy_path=tuple(mypypath),
-                       package_path=tuple(search_dirs),
-                       typeshed_path=tuple(lib_path))
+    return SearchPaths(
+        python_path=tuple(reversed(python_path)),
+        mypy_path=tuple(mypypath),
+        package_path=tuple(sys_path + site_packages),
+        typeshed_path=tuple(lib_path),
+    )
 
 
 def load_stdlib_py_versions(custom_typeshed_dir: Optional[str]) -> StdlibVersions:

--- a/mypy/pyinfo.py
+++ b/mypy/pyinfo.py
@@ -7,12 +7,13 @@ library found in Python 2. This file is run each mypy run, so it should be kept 
 possible.
 """
 import os
+import site
 import sys
 import sysconfig
 
 MYPY = False
 if MYPY:
-    from typing import List
+    from typing import Tuple, List
 
 if __name__ == '__main__':
     # HACK: We don't want to pick up mypy.types as the top-level types
@@ -24,7 +25,22 @@ if __name__ == '__main__':
     sys.path = old_sys_path
 
 
-def getsearchdirs():
+def getsitepackages():
+    # type: () -> List[str]
+    res = []
+    if hasattr(site, "getsitepackages"):
+        res.extend(site.getsitepackages())
+
+        if hasattr(site, "getusersitepackages") and site.ENABLE_USER_SITE:
+            res.insert(0, site.getusersitepackages())
+    else:
+        from distutils.sysconfig import get_python_lib
+
+        res = [get_python_lib()]
+    return res
+
+
+def getsyspath():
     # type: () -> List[str]
     # Do not include things from the standard library
     # because those should come from typeshed.
@@ -52,6 +68,14 @@ def getsearchdirs():
 
     abs_sys_path = (os.path.abspath(p) for p in sys.path[offset:])
     return [p for p in abs_sys_path if p not in excludes]
+
+
+def getsearchdirs():
+    # type: () -> Tuple[List[str], List[str]]
+    return (
+        getsyspath(),
+        getsitepackages(),
+    )
 
 
 if __name__ == '__main__':

